File: src/main/java/at/ac/uibk/dps/cirrina/execution/object/statemachine/StateMachineEventHandler.java
package at.ac.uibk.dps.cirrina.execution.object.statemachine;

import at.ac.uibk.dps.cirrina.execution.object.event.Event;
import at.ac.uibk.dps.cirrina.execution.object.event.EventHandler;
import at.ac.uibk.dps.cirrina.observability.tracing.MethodName;
import at.ac.uibk.dps.cirrina.observability.tracing.Trace;
import java.io.IOException;

public class StateMachineEventHandler {

  private final StateMachine stateMachine;

  private final EventHandler eventHandler;

  public StateMachineEventHandler(StateMachine stateMachine, EventHandler eventHandler) {
    this.stateMachine = stateMachine;
    this.eventHandler = eventHandler;
  }

  @Trace(name = MethodName.SEND_EVENT)
  public void sendEvent(Event event) throws IOException {
    eventHandler.sendEvent(event, stateMachine.getStateMachineInstanceId().toString());
  }
}

File: src/main/java/at/ac/uibk/dps/cirrina/execution/object/statemachine/StateMachineEventHandler.kt
package at.ac.uibk.dps.cirrina.execution.`object`.statemachine

import at.ac.uibk.dps.cirrina.execution.`object`.event.Event
import at.ac.uibk.dps.cirrina.execution.`object`.event.EventHandler

/**
 * A handler responsible for routing events from a [stateMachine] to an underlying [eventHandler].
 *
 * This class acts as a bridge, ensuring that events sent by the state machine are correctly
 * associated with the machine's unique identifier.
 *
 * @property stateMachine the state machine instance associated with this handler.
 * @property eventHandler the underlying handler used for dispatching events.
 */
class StateMachineEventHandler(
  private val stateMachine: StateMachine,
  private val eventHandler: EventHandler,
) {

  /**
   * Sends an [event] to the underlying event handler, tagged with the state machine's identifier.
   *
   * @param event the event to be sent.
   */
  fun sendEvent(event: Event) {
    eventHandler.sendEvent(event, stateMachine.id)
  }
}
